sdks/python/apache_beam/runners/direct/transform_evaluator.py

```diff
@@ -357,6 +361,85 @@ def finish_bundle(self):
         {None: hold})
 
 
+class _PubSubSubscriptionWrapper(object):
+  """Wrapper for garbage-collecting temporary PubSub subscriptions."""
+
+  def __init__(self, subscription, should_cleanup):
+    self.subscription = subscription
+    self.should_cleanup = should_cleanup
+
+  def __del__(self):
+    if self.should_cleanup:
+      self.subscription.delete()
+
+
+class _PubSubReadEvaluator(_TransformEvaluator):
+  """TransformEvaluator for PubSub read."""
+
+  _subscription_cache = {}
+
+  def __init__(self, evaluation_context, applied_ptransform,
+               input_committed_bundle, side_inputs, scoped_metrics_container):
+    assert not side_inputs
+    super(_PubSubReadEvaluator, self).__init__(
+        evaluation_context, applied_ptransform, input_committed_bundle,
+        side_inputs, scoped_metrics_container)
+
+    source = self._applied_ptransform.transform._source
+    should_create = not source.subscription_name
+    self._subscription = _PubSubReadEvaluator.get_subscription(
+        self._applied_ptransform, source.project, source.topic_name,
+        source.subscription_name, should_create)
+
+  @classmethod
+  def get_subscription(cls, transform, project, topic, subscription_name,
+                       should_create):
+    if transform not in cls._subscription_cache:
+      from google.cloud import pubsub
+      if should_create:
+        subscription_name = 'dataflow_%x' % random.randrange(1 << 32)
```

Comment:
Perhaps use 'beam' instead of dataflow. Also it might be better to have a timestamp in the subscription name (instead of or together with the random).